### PR TITLE
Add version annotation to various flags added in 1.13

### PR DESCRIPTION
**Pull request by thaJeztah**
Pull request https://github.com/docker/docker/pull/27745 added support for the
client to talk to older versions of the daemon. Various flags were added to
docker 1.13 that are not compatible with older daemons.

This PR adds annotations to those flags, so that they are automatically hidden
if the daemon is older than docker 1.13 (API 1.25).

Not all new flags affect the API (some are client-side only). The following
PR's added new flags to docker 1.13 that affect the API;

- https://github.com/docker/docker/pull/23430 added `--cpu-rt-period`and `--cpu-rt-runtime`
- https://github.com/docker/docker/pull/27800 / https://github.com/docker/docker/pull/25317 added `--group` / `--group-add` / `--group-rm`
- https://github.com/docker/docker/pull/27702 added `--network` to `docker build`
- https://github.com/docker/docker/pull/25962 added `--attachable` to `docker network create`
- https://github.com/docker/docker/pull/27998 added `--compose-file` to `docker stack deploy`
- https://github.com/docker/docker/pull/22566 added `--stop-timeout` to `docker run` and `docker create`
- https://github.com/docker/docker/pull/26061 added `--init` to `docker run` and `docker create`
- https://github.com/docker/docker/pull/26941 added `--init-path` to `docker run` and `docker create`
- https://github.com/docker/docker/pull/27958 added `--cpus` on `docker run` / `docker create`
- https://github.com/docker/docker/pull/27567 added `--dns`, `--dns-opt`, and `--dns-search` to `docker service create`
- https://github.com/docker/docker/pull/27596 added `--force` to `docker service update`
- https://github.com/docker/docker/pull/27857 added `--hostname` to `docker service create`
- https://github.com/docker/docker/pull/28031 added `--hosts`, `--host-add` / `--host-rm` to `docker service create` and `docker service update`
- https://github.com/docker/docker/pull/28076 added `--tty` on `docker service create` / `docker service update`
- https://github.com/docker/docker/pull/26421 added `--update-max-failure-ratio`, `--update-monitor` and `--rollback` on `docker service update`
- https://github.com/docker/docker/pull/27369 added `--health-cmd`, `--health-interval`, `--health-retries`, `--health-timeout` and `--no-healthcheck` options to `docker service create` and `docker service update`

**- How to verify it**

Verify that the correct flags are hidden when overriding the API version. For example;

Without `DOCKER_API_VERSION`;

```
docker service create --help

Usage:	docker service create [OPTIONS] IMAGE [COMMAND] [ARG...]

Create a new service

Options:
      --constraint list                  Placement constraints (default [])
      --container-label list             Container labels (default [])
      --dns list                         Set custom DNS servers (default [])
      --dns-option list                  Set DNS options (default [])
      --dns-search list                  Set custom DNS search domains (default [])
      --endpoint-mode string             Endpoint mode (vip or dnsrr)
  -e, --env list                         Set environment variables (default [])
      --env-file list                    Read in a file of environment variables (default [])
      --group list                       Set one or more supplementary user groups for the container (default [])
      --health-cmd string                Command to run to check health
      --health-interval duration         Time between running the check (ns|us|ms|s|m|h)
      --health-retries int               Consecutive failures needed to report unhealthy
      --health-timeout duration          Maximum time to allow one check to run (ns|us|ms|s|m|h)
      --help                             Print usage
      --host list                        Set one or more custom host-to-IP mappings (host:ip) (default [])
      --hostname string                  Container hostname
  -l, --label list                       Service labels (default [])
      --limit-cpu decimal                Limit CPUs (default 0.000)
      --limit-memory bytes               Limit Memory (default 0 B)
      --log-driver string                Logging driver for service
      --log-opt list                     Logging driver options (default [])
      --mode string                      Service mode (replicated or global) (default "replicated")
      --mount mount                      Attach a filesystem mount to the service
      --name string                      Service name
      --network list                     Network attachments (default [])
      --no-healthcheck                   Disable any container-specified HEALTHCHECK
  -p, --publish port                     Publish a port as a node port
      --replicas uint                    Number of tasks
      --reserve-cpu decimal              Reserve CPUs (default 0.000)
      --reserve-memory bytes             Reserve Memory (default 0 B)
      --restart-condition string         Restart when condition is met (none, on-failure, or any)
      --restart-delay duration           Delay between restart attempts (ns|us|ms|s|m|h)
      --restart-max-attempts uint        Maximum number of restarts before giving up
      --restart-window duration          Window used to evaluate the restart policy (ns|us|ms|s|m|h)
      --secret secret                    Specify secrets to expose to the service
      --stop-grace-period duration       Time to wait before force killing a container (ns|us|ms|s|m|h)
  -t, --tty                              Allocate a pseudo-TTY
      --update-delay duration            Delay between updates (ns|us|ms|s|m|h) (default 0s)
      --update-failure-action string     Action on update failure (pause|continue) (default "pause")
      --update-max-failure-ratio float   Failure rate to tolerate during an update
      --update-monitor duration          Duration after each task update to monitor for failure (ns|us|ms|s|m|h) (default 0s)
      --update-parallelism uint          Maximum number of tasks updated simultaneously (0 to update all at once) (default 1)
  -u, --user string                      Username or UID (format: <name|uid>[:<group|gid>])
      --with-registry-auth               Send registry authentication details to swarm agents
  -w, --workdir string                   Working directory inside the container
```  

And with `DOCKER_API_VERSION`;

```bash  
DOCKER_API_VERSION=1.24 docker service create --help

Usage:	docker service create [OPTIONS] IMAGE [COMMAND] [ARG...]

Create a new service

Options:
      --constraint list                Placement constraints (default [])
      --container-label list           Container labels (default [])
      --endpoint-mode string           Endpoint mode (vip or dnsrr)
  -e, --env list                       Set environment variables (default [])
      --env-file list                  Read in a file of environment variables (default [])
      --help                           Print usage
  -l, --label list                     Service labels (default [])
      --limit-cpu decimal              Limit CPUs (default 0.000)
      --limit-memory bytes             Limit Memory (default 0 B)
      --log-driver string              Logging driver for service
      --log-opt list                   Logging driver options (default [])
      --mode string                    Service mode (replicated or global) (default "replicated")
      --mount mount                    Attach a filesystem mount to the service
      --name string                    Service name
      --network list                   Network attachments (default [])
  -p, --publish port                   Publish a port as a node port
      --replicas uint                  Number of tasks
      --reserve-cpu decimal            Reserve CPUs (default 0.000)
      --reserve-memory bytes           Reserve Memory (default 0 B)
      --restart-condition string       Restart when condition is met (none, on-failure, or any)
      --restart-delay duration         Delay between restart attempts (ns|us|ms|s|m|h)
      --restart-max-attempts uint      Maximum number of restarts before giving up
      --restart-window duration        Window used to evaluate the restart policy (ns|us|ms|s|m|h)
      --stop-grace-period duration     Time to wait before force killing a container (ns|us|ms|s|m|h)
      --update-delay duration          Delay between updates (ns|us|ms|s|m|h) (default 0s)
      --update-failure-action string   Action on update failure (pause|continue) (default "pause")
      --update-parallelism uint        Maximum number of tasks updated simultaneously (0 to update all at once) (default 1)
  -u, --user string                    Username or UID (format: <name|uid>[:<group|gid>])
      --with-registry-auth             Send registry authentication details to swarm agents
  -w, --workdir string                 Working directory inside the container
```

**- Description for the changelog**

- Hide flags that are not supported by the daemon in `--help` output

### CODE REVIEW

File: cli/command/container/opts.go [hunk 1]
```diff
@@ -236,9 +236,12 @@ func addFlags(flags *pflag.FlagSet) *containerOptions {
 	flags.Int64Var(&copts.cpuPeriod, "cpu-period", 0, "Limit CPU CFS (Completely Fair Scheduler) period")
 	flags.Int64Var(&copts.cpuQuota, "cpu-quota", 0, "Limit CPU CFS (Completely Fair Scheduler) quota")
 	flags.Int64Var(&copts.cpuRealtimePeriod, "cpu-rt-period", 0, "Limit CPU real-time period in microseconds")
+	flags.SetAnnotation("cpu-rt-period", "version", []string{"1.25"})
```
Comment (thaJeztah) [inline cli/command/container/opts.go]: Wondering if we should add a utility for this, or to specify the version when adding flags (this does the job, but it's rather ugly to have these `SetAnnotation()` everywhere.

**Comment by thaJeztah** [comment]
ping @vieux @dnephin ptal

File: cli/command/container/opts.go (hunk 1 above)
Comment (dnephin) [inline cli/command/container/opts.go]: We could have a util and group the flags by version:

```go

SetFlagVersion(flags, "1.25", "cpu-rt-period", "cpu-rt-runtime", "cpus", "init", "init-path")

```

**Review by dnephin — APPROVED** [review]
LGTM

File: cli/command/container/opts.go (hunk 1 above)
Comment (thaJeztah) [inline cli/command/container/opts.go]: Yes I considered putting all the annotations at the end (looking a bit cleaner), but decided keeping them together with the flag also has it's advantages.

Ideally, something like this;

```go
flags.Var(&copts.foo, "foo", "Set foo option", flagAnnotation{"version", "1.25"},  flagAnnotation{"platform", "windows"})
```

or

```go
flags.Var(&copts.foo, "foo", "Set foo option").Annotate("version", "1.25").Annotate("platform", "windows")
```

Comment (dnephin) [inline cli/command/container/opts.go]: We'll have to fork `pflag` to make that kind of change, but that's fine with me. Having the flag Vars return the object would be a good idea.

However many of these lines are already too long, so I don't know that adding more to the line is an improvement.

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**